File: src/leetcode/python/medium/PartitionEqualSubsetSum_416.py
class PartitionEqualSubsetSum_416(object):
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        sm = sum(nums)
        if sm % 2 != 0:
            return False
        sm /= 2
        reachable = [False] * (sm + 1)
        reachable[0] = True
        for n in nums:
            for j in range(sm, n, -1):
                reachable[j] = reachable[j] or reachable[j - n]
        return reachable[sm]

File: src/leetcode/java/PartitionEqualSubsetSum_416.java
package leetcode.java;

public class PartitionEqualSubsetSum_416 {
  public boolean canPartition(int[] nums) {
    int sum = 0;
    for (int num : nums)
      sum += num;
    if (sum % 2 != 0) return false;
    sum /= 2;
    boolean[] reachable = new boolean[sum + 1];
    reachable[0] = true;
    for (int num : nums) {
      for (int j = sum; j >= num; j--) {
        reachable[j] = reachable[j] || reachable[j - num];
      }
    }
    return reachable[sum];
  }
}
